Add LDC-specific traits for CTFE information about the target machine.

JohanEngelen opened the pull request:
Edit: updated to current state of the PR.  (removed `__traits(targetFeatures)`)

This adds the following LDC-specific traits:
```D
__traits(targetCPU) == "broadwell"
__traits(targetHasFeature, "sse2") == bool
```

Example usage:
```D
import std.stdio;
void main() {
  writeln("CPU = ", __traits(targetCPU));
  writeln("Has 'sse3' = ", __traits(targetHasFeature, "sse3"));
  writeln("Has 'sse4' = ", __traits(targetHasFeature, "sse4"));
  writeln("Has 'sse4.1' = ", __traits(targetHasFeature, "sse4.1"));
}
```
Outputs on my machine:
```
❯ ../bin/ldc2 -wi -mcpu=native -run ../../tests/codegen/target_traits.d
CPU = haswell
Has 'sse3' = true
Has 'sse4' = false
Has 'sse4.1' = true
```

JohanEngelen commented:
@9il what do you think?

9il commented:
Awesome! Thank you a lot!

WalterBright commented:
```
__traits(targetFeatures) == Tuple("sse", "sse2", ...)
__traits(targetHasFeature, "sse2") == bool
```
The latter is redundant. We should try to avoid redundant core language features.

UplinkCoder commented:
so is hasMember

JohanEngelen commented:
> The latter is redundant. We should try to avoid redundant core language features.

I left targetHasFeature for convenience.
`any!q{a = "sse4.1"}([__traits(targetFeatures)])` is quite ugly I think for functionality that is probably most used out of the three traits.

ibuclaw commented:
I can't say that gcc readily provides this information, but it seems reasonable.

WalterBright commented:
> so is hasMember

True, but we learn from our mistakes rather than repeat them.

> is quite ugly

`__traits` is always ugly. It was intended to be buried in library code. Things that are easily done in the library should not be in the core language.

JohanEngelen commented:
> I can't say that gcc readily provides this information

Neither does LLVM :(
For -mcpu=native the information is easy to obtain, for -mcpu=haswell it is not...

ibuclaw commented:
> Neither does LLVM :(

Well at least you have `getTargetFeatureString` and `getTargetCPU` exposed to the frontend. ;-)

JohanEngelen commented:
OK, I've been able to squeeze the feature info out of LLVM with a small LLVM patch :)
I'll submit that patch to LLVM, and see how that goes over. Then for LLVM<3.9, __traits(targetHasFeature, ...) will only work when either explicitly mentioning the feature on the cmdline, or with cpu=native.

ibuclaw commented:
I guess you could cache this information in `Target` also.  That may be the best way for me to do it without adding a new gcc target "hook".

JohanEngelen commented:
http://reviews.llvm.org/D19189

mleise commented:
Nice! I hope this makes it into all compilers. There is always the option to define a safe fall-back like "i386" for dmd on x86. `targetHasFeature` would be the most frequently used trait in the set. The list version is really only good for "compiled with \<features\>" output ... unless someone writes a nice wrapper.

@ibuclaw Maybe checking the implementation of this helps:
```
gcc -march=native -dM -E - < /dev/null | egrep "MMX|SSE|AVX" | sort
#define __AVX__ 1
#define __AVX2__ 1
#define __MMX__ 1
#define __SSE__ 1
…
```

ibuclaw commented:
@mleise - I'm aware of how gcc generates macros, gdc uses the same mechanism for generating pre-defined version conditions. ;-)

All I said was that there is no "ready to use" way of getting the information.

JohanEngelen commented:
I removed `__traits(targetFeatures)` as this list is not obtainable without hacking LLVM. `__traits(targetHasFeature)` is more useful and this can be obtained from LLVM with some effort.
Updated the PR.

JohanEngelen commented:
For LLVM version < 3.7, `hasTargetFeature` will always return the safe `false` default. I don't know how to obtain the necessary information with LLVM < 3.7.

JohanEngelen commented:
Ping! :)

9il commented:
LGTM

This would be very helpful for [Mir's BLAS Micro Kernel](https://github.com/libmir/mir/pull/210/files) parameter estimation. LDC can generates [(see gist)](https://gist.github.com/9il/58c1b80110de2db5f2eff6999346a928) efficient micro kernel code, and #1472 will allow to optimize it even more :-)

JohanEngelen commented:
@redstar Can I merge this?  The "hook" is the same as #5771 so it is easily modified once that is merged.

JohanEngelen commented:
Will merge when green.

JohanEngelen commented:
@9il Can you update us on how you are using this in your code? :)

9il commented:
Will do with BLAS release :)
